Improve the package image fusion library.

RichardTamarisk opened the pull request:
* Remove unnecessary comments in task.hpp and stitcher.cpp.
* Update main(): The video frames decoded by ffmpeg are spliced.

RichardTamarisk committed:
* Remove unnecessary comments in task.hpp and stitcher.cpp.
* Update main(): The video frames decoded by ffmpeg are spliced.

RichardTamarisk committed:
* Add stitcher.cpp and stitcher.h to src.
* update CMakeLists.txt.

RichardTamarisk commented:
In this change, I add the image stitching wrapped function implementation stitcher.cpp, and successfully compiled the project, but the nothing in the SDL window after running the project.
Usage:
* ./DisplayImage ../../fusion_fuc/img/left_1.jpg ../../fusion_fuc/img/right_1.jpg
feedback:
Task init! 0x561a80352cd0
No more packet!
No more packet!
* ./DisplayImage ../img/left.mp4 ../img/right.mp4
feedback:
Task init! 0x644d7cc77cd0
Segmentation fault (core dumped)

I was hoping you could take a look at this.

JackLau1222 commented:
> In this change, I add the image stitching wrapped function implementation stitcher.cpp, and successfully compiled the project, but the nothing in the SDL window after running the project. Usage:
> 
> * ./DisplayImage ../../fusion_fuc/img/left_1.jpg ../../fusion_fuc/img/right_1.jpg
>   feedback:
>   Task init! 0x561a80352cd0
>   No more packet!
>   No more packet!
> * ./DisplayImage ../img/left.mp4 ../img/right.mp4
>   feedback:
>   Task init! 0x644d7cc77cd0
>   Segmentation fault (core dumped)
> 
> I was hoping you could take a look at this.

I reviewed your code and have the following suggestions:

Please set is_corrected to false in the ImageFusion function, as the correct function is not robust enough. It doesn't handle some edge cases well, which leads to errors.
Also, do you really need to convert to the AV_PIX_FMT_BGR24 color format? On my Mac platform, this conversion is not accelerated, which causes errors. You might want to try it on your platform to see if it works properly.

The error log like this
```bash
[swscaler @ 0x1381f8000] No accelerated colorspace conversion found from yuv420p to bgr24.
[swscaler @ 0x1381f8000] bad src image pointers
sws_scale failed.
One or both input images are empty.
```

RichardTamarisk committed:
* Add test: feature point number.

RichardTamarisk commented:
The feature point detector I used before indicated that AV_PIX_FMT_BGR24 color space was required, so I changed to another detector. In my fusion_fuc, I conducted code test without any problem, whether is_correct is false or true is no problem.Usage:
```
cd ImageFusion/code/fusion_fuc
mkdir build
cd build
cmake ..
make
./DisplayImage ../img/left_1.jpg ../img/right_1.jpg
```
In addition, I also checked the number of detected features is normal.
<img width="972" alt="image" src="https://github.com/user-attachments/assets/284eee5d-95cd-4e8a-8f83-dfeeeda9c9af">

JackLau1222 commented:
> The feature point detector I used before indicated that AV_PIX_FMT_BGR24 color space was required, so I changed to another detector. In my fusion_fuc, I conducted code test without any problem, whether is_correct is false or true is no problem.Usage:
> 
> ```
> cd ImageFusion/code/fusion_fuc
> mkdir build
> cd build
> cmake ..
> make
> ./DisplayImage ../img/left_1.jpg ../img/right_1.jpg
> ```
> 
> In addition, I also checked the number of detected features is normal. <img alt="image" width="972" src="https://private-user-images.githubusercontent.com/128676569/390935718-284eee5d-95cd-4e8a-8f83-dfeeeda9c9af.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzI4NDQ0MTIsIm5iZiI6MTczMjg0NDExMiwicGF0aCI6Ii8xMjg2NzY1NjkvMzkwOTM1NzE4LTI4NGVlZTVkLTk1Y2QtNGU4YS04ZjgzLWRmZWVlZGE5YzlhZi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMTI5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTEyOVQwMTM1MTJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0wYzU3NTE5ODUzYjU0NGIxZmVjMGZlOWM0NjY3NGFiMGUxNjNmNzg1ZTZiYjc2YjEzYjdhMjBkNjFmODU3ZjI0JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.r49pEv9Z3U8gg40iFZTYDM-7zYP8pB_PIzhGKkiQLgk">

you don't understand what i mean. i don't want to waste time to argue the code design here. it's not the point of this problem. 

first, i need you disable the correct in player.
second, i need you test the code in your platform to check if it work?

JackLau1222 commented:
what i want to tell you is the demo is not represent anything, you just pass simple scenario. the code can't handle any situation. 

if you don't understand this, the project will be no useful.

RichardTamarisk commented:
> The feature point detector I used before indicated that AV_PIX_FMT_BGR24 color space was required, so I changed to another detector. In my fusion_fuc, I conducted code test without any problem, whether is_correct is false or true is no problem.Usage:
> 
> ```
> cd ImageFusion/code/fusion_fuc
> mkdir build
> cd build
> cmake ..
> make
> ./DisplayImage ../img/left_1.jpg ../img/right_1.jpg
> ```
> 
> In addition, I also checked the number of detected features is normal. <img alt="image" width="972" src="https://private-user-images.githubusercontent.com/128676569/390935718-284eee5d-95cd-4e8a-8f83-dfeeeda9c9af.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzI4NDQ0MDcsIm5iZiI6MTczMjg0NDEwNywicGF0aCI6Ii8xMjg2NzY1NjkvMzkwOTM1NzE4LTI4NGVlZTVkLTk1Y2QtNGU4YS04ZjgzLWRmZWVlZGE5YzlhZi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMTI5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTEyOVQwMTM1MDdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1lMzkxNGE3NTljNDFjZTQwNGY4ODlhMjc1M2YwMzc4NzlkODI5NmExZmE2MGRjOTM3YjQ1MTE4MWZhOWE5YTZlJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.Zi4bahWna-7tobwjOXsL0Gos8VjvM9VZd7p5RsILsxY">



> > The feature point detector I used before indicated that AV_PIX_FMT_BGR24 color space was required, so I changed to another detector. In my fusion_fuc, I conducted code test without any problem, whether is_correct is false or true is no problem.Usage:
> > ```
> > cd ImageFusion/code/fusion_fuc
> > mkdir build
> > cd build
> > cmake ..
> > make
> > ./DisplayImage ../img/left_1.jpg ../img/right_1.jpg
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > In addition, I also checked the number of detected features is normal. <img alt="image" width="972" src="https://private-user-images.githubusercontent.com/128676569/390935718-284eee5d-95cd-4e8a-8f83-dfeeeda9c9af.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzI4NDQ0MTIsIm5iZiI6MTczMjg0NDExMiwicGF0aCI6Ii8xMjg2NzY1NjkvMzkwOTM1NzE4LTI4NGVlZTVkLTk1Y2QtNGU4YS04ZjgzLWRmZWVlZGE5YzlhZi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMTI5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTEyOVQwMTM1MTJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0wYzU3NTE5ODUzYjU0NGIxZmVjMGZlOWM0NjY3NGFiMGUxNjNmNzg1ZTZiYjc2YjEzYjdhMjBkNjFmODU3ZjI0JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.r49pEv9Z3U8gg40iFZTYDM-7zYP8pB_PIzhGKkiQLgk">
> 
> you don't understand what i mean. i don't want to waste time to argue the code design here. it's not the point of this problem.
> 
> first, i need you disable the correct in player. second, i need you test the code in your platform to check if it work?

I am currently testing whether the code can be played normally is video_fusion_player code, and the errors we encountered are completely different, I have not encountered the problem that the image after feature point detection and conversion image is empty, I hope you can upload your current code.

RichardTamarisk commented:
> what i want to tell you is the demo is not represent anything, you just pass simple scenario. the code can't handle any situation.
> 
> if you don't understand this, the project will be no useful.

I'll solve this problem as soon as possible.

JackLau1222 commented:
what i modified is just disable the correct_image. Just like what i commented
> I reviewed your code and have the following suggestions:

> Please set is_corrected to false in the ImageFusion function, as the correct function is not robust enough. It doesn't handle some edge cases well, which leads to errors.
> Also, do you really need to convert to the AV_PIX_FMT_BGR24 color format? On my Mac platform, this conversion is not accelerated, which causes errors. You might want to try it on your platform to see if it works properly.

JackLau1222 commented:
if you set is_corrected to false, you'll see what problem i encountered.

RichardTamarisk commented:
> > In this change, I add the image stitching wrapped function implementation stitcher.cpp, and successfully compiled the project, but the nothing in the SDL window after running the project. Usage:
> > 
> > * ./DisplayImage ../../fusion_fuc/img/left_1.jpg ../../fusion_fuc/img/right_1.jpg
> >   feedback:
> >   Task init! 0x561a80352cd0
> >   No more packet!
> >   No more packet!
> > * ./DisplayImage ../img/left.mp4 ../img/right.mp4
> >   feedback:
> >   Task init! 0x644d7cc77cd0
> >   Segmentation fault (core dumped)
> > 
> > I was hoping you could take a look at this.
> 
> I reviewed your code and have the following suggestions:
> 
> Please set is_corrected to false in the ImageFusion function, as the correct function is not robust enough. It doesn't handle some edge cases well, which leads to errors. Also, do you really need to convert to the AV_PIX_FMT_BGR24 color format? On my Mac platform, this conversion is not accelerated, which causes errors. You might want to try it on your platform to see if it works properly.
> 
> The error log like this
> 
> ```shell
> [swscaler @ 0x1381f8000] No accelerated colorspace conversion found from yuv420p to bgr24.
> [swscaler @ 0x1381f8000] bad src image pointers
> sws_scale failed.
> One or both input images are empty.
> ```

Yes, in image fusion processing, I need to convert to BGR24 to process the tree color channels.
I am currently working on converting the image to YUV420P after the image stitching process is finished.
Maybe you don't care what  needs to be done in this process, but I wonder if it needs to be done this way?

JackLau1222 commented on code/video_fusion_player/src/stitcher.cpp:
```diff
@@ -0,0 +1,387 @@
+    }
+
+    // 使用 sws_scale 转换图像
+    const uint8_t *srcSlice[1] = { frame->data[0] }; 
```
There is one problem obviously. The AVFrame data maybe stored in multiple planes. For example, YUV420p has Y, U, and V.

RichardTamarisk committed:
* update function: avframeToCvmat and cvmatToAvframe.

RichardTamarisk commented:
I improved the avframeToCvmat function for submission this time, and the converted cv::Mat format images can be splintered normally, but the AVFrame format images converted by cvmatToAvframe function are splintered. Could you please help me find out what is wrong.

RichardTamarisk commented:
The resulting image is spliced after using avframeToCvmat:
![image](https://github.com/user-attachments/assets/8378b4c4-2661-4ce4-b104-8714cdf523d5)
Use cvmatToAvframe to convert the spliced image to avframe:
![image](https://github.com/user-attachments/assets/53e3cef9-6fc9-43a4-98d1-658b27a08467)
